File: src/Core/MeshPong.fs
namespace Zeta.Core

open System.Threading.Tasks

/// MeshPong — **pong over the mesh: the first playable integration artifact** (the demo named in the
/// game-physics-over-Reticulum doc).
///
/// The **lockstep model** (Age of Empires "1500 archers" / GGPO's core trick), on our substrate: **only
/// INPUTS travel as membrane crossings; the seed carries the world.** Two rooms each replay the SAME
/// recorded input crossings (`RecordedSource`) and independently simulate the WHOLE game. Because the
/// physics is a pure **integer** fold (no floats ⇒ byte-stable), their final states must agree exactly —
/// **agreement = the treaty holding; divergence = desync = a treaty violation, detected by comparison**.
/// A tampered crossing is caught the same way: **anti-cheat = noninterference, demonstrated in a test.**
///
/// This demo also EARNED a substrate improvement: lockstep needs the crossing's PAYLOAD, which exposed
/// that `SoftScheduler.Handler` never received the arrival — hence the additive `HandlerK`/`driveK` +
/// `SimFramework.RoomK` (arrival-aware rooms). Pure module; deterministic; DST throughout. The integer
/// game is portable to the other three oracles (a future game-state treaty).
[<RequireQualifiedAccess>]
module MeshPong =

    [<Literal>]
    let Width = 16

    [<Literal>]
    let Height = 9

    [<Literal>]
    let PaddleLen = 3

    /// The whole game — small, integer, byte-stable.
    type Game =
        { BallX: int
          BallY: int
          VX: int
          VY: int
          PaddleA: int // top row of paddle A (left wall)
          PaddleB: int // top row of paddle B (right wall)
          ScoreA: int
          ScoreB: int }

    /// Center serve, heading right-down.
    let create () : Game =
        { BallX = Width / 2
          BallY = Height / 2
          VX = 1
          VY = 1
          PaddleA = Height / 2 - 1
          PaddleB = Height / 2 - 1
          ScoreA = 0
          ScoreB = 0 }

    let private clampPaddle (p: int) = max 0 (min (Height - PaddleLen) p)

    let private serve (vx: int) (g: Game) =
        { g with BallX = Width / 2; BallY = Height / 2; VX = vx; VY = 1 }

    /// One lockstep tick: both players' inputs (each clamped to {-1,0,+1}) + the pure physics fold.
    let step (inputA: int) (inputB: int) (g: Game) : Game =
        let g =
            { g with
                PaddleA = clampPaddle (g.PaddleA + max -1 (min 1 inputA))
                PaddleB = clampPaddle (g.PaddleB + max -1 (min 1 inputB)) }

        let nx = g.BallX + g.VX
        let nyRaw = g.BallY + g.VY
        // bounce top/bottom
        let vy, ny =
            if nyRaw < 0 || nyRaw >= Height then -g.VY, g.BallY - g.VY else g.VY, nyRaw

        if nx <= 0 then
            if ny >= g.PaddleA && ny < g.PaddleA + PaddleLen then
                { g with BallX = 1; BallY = ny; VX = 1; VY = vy } // A saves
            else
                serve -1 { g with ScoreB = g.ScoreB + 1 } // B scores; serve toward A
        elif nx >= Width - 1 then
            if ny >= g.PaddleB && ny < g.PaddleB + PaddleLen then
                { g with BallX = Width - 2; BallY = ny; VX = -1; VY = vy } // B saves
            else
                serve 1 { g with ScoreA = g.ScoreA + 1 } // A scores; serve toward B
        else
            { g with BallX = nx; BallY = ny; VY = vy }

    // ── The mesh wiring: inputs are the ONLY crossings ──

    /// Deterministic input pair for a tick (each ∈ {-1,0,+1}) — the stand-in "controllers"; with real
    /// players the wire shape is identical (a Reticulum packet carrying the same payload).
    let inputsAt (seed: int64) (tick: int) : int * int =
        let h = SplitMix64.mix (uint64 seed + uint64 tick * SplitMix64.GoldenRatio)
        (int (h % 3UL)) - 1, (int ((h >>> 8) % 3UL)) - 1

    /// Encode a tick's input pair as the crossing payload (text — the treaty register).
    let encodeInputs (a: int) (b: int) : string = sprintf "pong:%d,%d" a b

    /// Parse a crossing payload back to the input pair (None = not a pong input — honest refusal).
    let parseInputs (payload: string) : (int * int) option =
        if payload.StartsWith "pong:" then
            match payload.Substring(5).Split(',') with
            | [| a; b |] ->
                match System.Int32.TryParse a, System.Int32.TryParse b with
                | (true, ia), (true, ib) -> Some(ia, ib)
                | _ -> None
            | _ -> None
        else
            None

    /// The "controllers" membrane: each tick, both players' inputs cross as ONE message.
    let inputsSource (seed: int64) : SoftScheduler.Source =
        fun tick ->
            let a, b = inputsAt seed tick
            [ OperatorMessageArrived(encodeInputs a b) ]

    /// The lockstep handler — arrival-AWARE: reads the crossing's payload, steps the pure physics.
    /// A non-pong payload is skipped (state unchanged) — the membrane may carry other traffic.
    let lockstepHandler: SoftScheduler.HandlerK<Game> =
        SoftScheduler.handlerK
            "pong-lockstep"
            (function
            | OperatorMessageArrived _ -> true
            | _ -> false)
            (fun intr _ctx g ->
                match intr with
                | OperatorMessageArrived payload ->
                    match parseInputs payload with
                    | Some (a, b) -> Task.FromResult(Ok(step a b g))
                    | None -> Task.FromResult(Ok g)
                | _ -> Task.FromResult(Ok g))

    /// A pong room over a given membrane. Resolves (signs off) when anyone scores `toWin` points.
    let room (name: string) (source: int64 -> SoftScheduler.Source) (budget: int) (toWin: int) : SimFramework.RoomK<Game> =
        { Name = name
          Initial = fun _ -> create ()
          HandlersK = [ lockstepHandler ]
          Source = source
          Budget = budget
          Resolved = fun g -> g.ScoreA >= toWin || g.ScoreB >= toWin }

    /// **Play a match over the mesh**: record the input crossings once (the "network session"), then TWO
    /// rooms each replay the same recording and simulate independently. Returns (recording, reportA,
    /// reportB) — the lockstep treaty check is `reportA.Final = reportB.Final` (byte-equal worlds).
    let playMatch (seed: int64) (budget: int) (toWin: int) =
        task {
            let recording = RecordedSource.record (inputsSource seed) budget
            let replay = fun (_: int64) -> RecordedSource.replay recording
            let! reportA = SimFramework.runK (room "pong-room-A" replay budget toWin) seed
            let! reportB = SimFramework.runK (room "pong-room-B" replay budget toWin) seed
            return recording, reportA, reportB
        }

    // ── The GAME-STATE TREATY codec (Aaron 2026-06-11: "yes on the game treaty") ──
    // Canonical text line for the Game state; all four oracles replay the shared session
    // (src/Core.TypeScript/mesh-pong/golden-vectors.lines) through their own pure `step` and must hit
    // byte-identical checkpoint lines. Integer fields, tab-separated, versioned tag.

    /// Serialize the game state to its canonical treaty line (byte-deterministic).
    let gameToLine (g: Game) : string =
        sprintf
            "ponggame1\t%d\t%d\t%d\t%d\t%d\t%d\t%d\t%d"
            g.BallX
            g.BallY
            g.VX
            g.VY
            g.PaddleA
            g.PaddleB
            g.ScoreA
            g.ScoreB

    /// Parse a canonical game-state line (None = malformed — honest refusal).
    let gameOfLine (line: string) : Game option =
        match line.Split('\t') with
        | [| "ponggame1"; bx; by; vx; vy; pa; pb; sa; sb |] ->
            let p (s: string) =
                match System.Int32.TryParse s with
                | true, v -> Some v
                | _ -> None
            match p bx, p by, p vx, p vy, p pa, p pb, p sa, p sb with
            | Some bx, Some by, Some vx, Some vy, Some pa, Some pb, Some sa, Some sb ->
                Some
                    { BallX = bx
                      BallY = by
                      VX = vx
                      VY = vy
                      PaddleA = pa
                      PaddleB = pb
                      ScoreA = sa
                      ScoreB = sb }
            | _ -> None
        | _ -> None

File: src/Core.TypeScript/mesh-pong/mesh-pong.js
/**
 * mesh-pong — the mesh-pong world (integer-grid lockstep pong). TS parity oracle; mirrors
 * src/Core/MeshPong.fs (the F# oracle that LOCKED the game-state treaty) and the C#/Rust siblings.
 * GAME-STATE TREATY: step + toLine must replay the shared session (./golden-vectors.lines) to
 * byte-identical checkpoint lines.
 */
export const WIDTH = 16;
export const HEIGHT = 9;
export const PADDLE_LEN = 3;
/** Center serve, heading right-down. */
export function create() {
    return {
        ballX: Math.floor(WIDTH / 2),
        ballY: Math.floor(HEIGHT / 2),
        vx: 1,
        vy: 1,
        paddleA: Math.floor(HEIGHT / 2) - 1,
        paddleB: Math.floor(HEIGHT / 2) - 1,
        scoreA: 0,
        scoreB: 0,
    };
}
function clampPaddle(p) {
    return Math.max(0, Math.min(HEIGHT - PADDLE_LEN, p));
}
function serve(g, vx) {
    return { ...g, ballX: Math.floor(WIDTH / 2), ballY: Math.floor(HEIGHT / 2), vx, vy: 1 };
}
/** One lockstep tick — the pure physics fold (parity with MeshPong.step). */
export function step(inputA, inputB, g0) {
    const g = {
        ...g0,
        paddleA: clampPaddle(g0.paddleA + Math.max(-1, Math.min(1, inputA))),
        paddleB: clampPaddle(g0.paddleB + Math.max(-1, Math.min(1, inputB))),
    };
    const nx = g.ballX + g.vx;
    const nyRaw = g.ballY + g.vy;
    const bounced = nyRaw < 0 || nyRaw >= HEIGHT;
    const vy = bounced ? -g.vy : g.vy;
    const ny = bounced ? g.ballY - g.vy : nyRaw;
    if (nx <= 0) {
        return ny >= g.paddleA && ny < g.paddleA + PADDLE_LEN
            ? { ...g, ballX: 1, ballY: ny, vx: 1, vy }
            : serve({ ...g, scoreB: g.scoreB + 1 }, -1);
    }
    if (nx >= WIDTH - 1) {
        return ny >= g.paddleB && ny < g.paddleB + PADDLE_LEN
            ? { ...g, ballX: WIDTH - 2, ballY: ny, vx: -1, vy }
            : serve({ ...g, scoreA: g.scoreA + 1 }, 1);
    }
    return { ...g, ballX: nx, ballY: ny, vy };
}
/** Parse a crossing payload ("pong:a,b"); null = not a pong input (honest refusal). */
export function parseInputs(payload) {
    if (!payload.startsWith("pong:"))
        return null;
    const parts = payload.slice(5).split(",");
    if (parts.length !== 2)
        return null;
    const a = Number(parts[0]);
    const b = Number(parts[1]);
    if (!Number.isInteger(a) || !Number.isInteger(b) || parts[0] === "" || parts[1] === "")
        return null;
    return [a, b];
}
/** Serialize to the canonical treaty line (byte-identical to the F# oracle). */
export function toLine(g) {
    return `ponggame1\t${g.ballX}\t${g.ballY}\t${g.vx}\t${g.vy}\t${g.paddleA}\t${g.paddleB}\t${g.scoreA}\t${g.scoreB}`;
}
/** Parse a canonical game-state line; null on malformed (honest refusal). */
export function ofLine(line) {
    const parts = line.split("\t");
    if (parts.length !== 9 || parts[0] !== "ponggame1")
        return null;
    const vals = [];
    for (let i = 1; i < 9; i++) {
        const v = Number(parts[i]);
        if (!Number.isInteger(v) || parts[i] === "")
            return null;
        vals.push(v);
    }
    return {
        ballX: vals[0],
        ballY: vals[1],
        vx: vals[2],
        vy: vals[3],
        paddleA: vals[4],
        paddleB: vals[5],
        scoreA: vals[6],
        scoreB: vals[7],
    };
}
